src/ansiblelint/rules/VariableNamingRule.py

```diff
@@ -1,74 +1,81 @@
-from typing import TYPE_CHECKING, List, Any, Dict, Union
-if TYPE_CHECKING:
-    from ansiblelint.constants import odict
-from ansiblelint.errors import MatchError
+import re
+from typing import TYPE_CHECKING, Any, Dict, Generator, List, Tuple, Union
+
 from ansiblelint.file_utils import Lintable
 from ansiblelint.rules import AnsibleLintRule
 from ansiblelint.utils import parse_yaml_from_file
-import ansible.parsing.yaml.objects
 
-import re
+if TYPE_CHECKING:
+    from ansiblelint.constants import odict
+    from ansiblelint.errors import MatchError
 
 
 # properties/parameters are prefixed and postfixed with `__`
-def is_property(k):
-    return (k.startswith('__') and k.endswith('__'))
+def is_property(k: str) -> bool:
+    """Check if key is a property."""
+    return k.startswith('__') and k.endswith('__')
 
 
-def is_invalid_variable_name(text):
+def is_invalid_variable_name(text: str) -> bool:
+    """Check if variable name is using right pattern."""
     patterns = '^[a-z0-9_]*$'
     if re.search(patterns, text):
         return False  # string uses acceptable characters
-    else:
-        return True  # string uses unacceptable characters
+    return True  # string uses unacceptable characters
 
 
 class VariableNamingRule(AnsibleLintRule):
     id = 'var-naming'
     base_msg = 'All variables should be named using only lowercase and underscores'
     shortdesc = base_msg
     description = 'All variables should be named using only lowercase and underscores'
-    severity = 'MEDIUM'  # ansible-lint displays severity when with --parseable-severity option
-    tags = ['formatting', 'readability']
-    version_added = 'v5.0.5'
-
-    def recursive_items(self, dictionary):
-        """Return a recursive search for all keys in the dictionary """
-
+    severity = (
+        'MEDIUM'  # ansible-lint displays severity when with --parseable-severity option
+    )
+    tags = ['formatting', 'readability', 'experimental']
+    version_added = 'v5.0.7'
+
+    def recursive_items(
+        self, dictionary: Dict[str, Any]
+    ) -> Generator[Tuple[str, Any], None, None]:
+        """Return a recursive search for all keys in the dictionary."""
         for key, value in dictionary.items():
             # Avoid internal properties in the dictionary
             if not is_property(key):
                 # Recurse if value is another ansible dictionary
-                if isinstance(value, ansible.parsing.yaml.objects.AnsibleMapping):
```

Comment:
In my testing this needed to be `if isinstance(value, ansible.parsing.yaml.objects.AnsibleMapping):` because checking for `dict` didn't work.

Reply:
Actually looks to be working - I just tested again. I think it was the addition of `yield (key, value)` that I added some time ago in my code that did it then. Closing this.